### PR TITLE
Feature/symbolic explanations

### PR DESCRIPTION
**Pull request by sarah-segel**
## Plugins
- Add symbolic explanations plugin (#46).

## Enhancements
- Update author email.
- Add an exit button that first deletes running jobs and then terminates DeepCave (otherwise jobs such as the symbolic regression would continue running after DeepCave termination).
- Nicer handling of Keyboard Interrupt (avoid long error message in terminal).
- Disable debug mode (to avoid error messages being shown directly in the UI, especially upon exit of DeepCave).

## Bug-Fixes
- Fix varying configspace index order in PDP and symbolic explanations.

## Others
- Disabled existing loggers again to suppress redis logs

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,7 +2,7 @@
 # are usually completed in github actions.
 
 SHELL := /bin/bash
-VERSION := 1.1.3
+VERSION := 1.2
```
Comment (helegraf) [inline Makefile:5]: This says version 1.2, but in the Changelog the change is under Version 1.1.4

File: deepcave/__init__.py [hunk 2]
```diff
@@ -9,15 +9,15 @@
...
+author_email = "s.segel@ai.uni-hannover.de"
 description = "An interactive framework to visualize and analyze your AutoML process in real-time."
 url = "automl.org"
 project_urls = {
     "Documentation": "https://automl.github.io/DeepCAVE/main",
     "Source Code": "https://github.com/automl/deepcave",
 }
 copyright = f"Copyright {datetime.date.today().strftime('%Y')}, {author}"
-version = "1.1.3"
+version = "1.2"
```
Comment (helegraf) [inline deepcave/__init__.py:20]: Same as the previous comment about inconsistent versions

File: docs/plugins/symbolic_explanations.rst [hunk 3]
```diff
@@ -0,0 +1,17 @@
+Symbolic Explanations
+====================
+
+This plugin is capable of answering similar questions as the Partial Dependencies plugin, i.e.:
+
+* How does the objective change wrt one or two hyperparameters? For example, does the accuracy
```
Comment (helegraf) [inline docs/plugins/symbolic_explanations.rst:6]: I would just write "with respect to"

File: docs/plugins/symbolic_explanations.rst [hunk 4]
```diff
@@ -0,0 +1,17 @@
+Symbolic Explanations
+====================
+
+This plugin is capable of answering similar questions as the Partial Dependencies plugin, i.e.:
```
Comment (helegraf) [inline docs/plugins/symbolic_explanations.rst:4]: I personally think an explanation of what symbolic explanations are in the first sentence would be best; so just adding one more sentence before this one.

File: deepcave/cli.py [hunk 5]
```diff
@@ -1,5 +1,4 @@
-from typing import Any, List
-
+import os
```
Comment (helegraf) [inline deepcave/cli.py:1]: os is unsused

File: deepcave/config.py [hunk 6]
```diff
@@ -49,6 +49,9 @@ def PLUGINS(self) -> Dict[str, List["Plugin"]]:
         from deepcave.plugins.budget.budget_correlation import BudgetCorrelation
         from deepcave.plugins.hyperparameter.importances import Importances
         from deepcave.plugins.hyperparameter.pdp import PartialDependencies
+        from deepcave.plugins.hyperparameter.symbolic_explanations import (
+            SymbolicExplanations,
+        )
```
Comment (helegraf) [inline deepcave/config.py:54]: Why the brackets? :)

File: deepcave/utils/symbolic_regression.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            raise ValueError(
+                f"Numer of hyperparameters to be explained by symbolic explanations must not "
+                f"be larger than 2"
+            )
+
+    # Simplify the formula
+    try:
+        # Simplification can fail in some cases. If so, use the unsimplified version.
+        symb_simpl = sympy.simplify(symb_conv)
+    except:
```
Comment (helegraf) [inline deepcave/utils/symbolic_regression.py:100]: Is there a specific exception that could be caught here instead?

File: deepcave/utils/symbolic_regression.py [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    exp_func = make_function(function=exp, arity=1, name="exp")
+
+    function_set = ["add", "sub", "mul", "div", "sqrt", "log", "sin", "cos", "abs", exp_func]
+
+    return function_set
+
+
+def convert_symb(symb, n_decimals: int = None, hp_names: list = None) -> sympy.core.expr:
+    """
+    Convert a fitted symbolic regression to a simplified and potentially rounded mathematical expression.
```
Comment (helegraf) [inline deepcave/utils/symbolic_regression.py:28]: This line is over 100 characters; however, I am not 100% sure what our desired setting for the linter would be. 100 is admittedly really short.

File: deepcave/utils/symbolic_regression.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+def get_function_set():
+    exp_func = make_function(function=exp, arity=1, name="exp")
+
+    function_set = ["add", "sub", "mul", "div", "sqrt", "log", "sin", "cos", "abs", exp_func]
+
+    return function_set
+
+
+def convert_symb(symb, n_decimals: int = None, hp_names: list = None) -> sympy.core.expr:
```
Comment (helegraf) [inline deepcave/utils/symbolic_regression.py:26]: Type of symb?

File: deepcave/plugins/hyperparameter/symbolic_explanations.py [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+            ),
+            dbc.Row(
+                [
+                    dbc.Col(
+                        [
+                            html.Div(
+                                [
+                                    dbc.Label("Parsimony coefficient"),
+                                    help_button(
+                                        "Penalizes the complexity of the resulting formulas."
```
Comment (helegraf) [inline deepcave/plugins/hyperparameter/symbolic_explanations.py:100]: Maybe add a line if higher or lower will mean less complexity

**Review by helegraf — COMMENTED** [review]
Looks nice overall and the tests are working 💪 I have some minor remarks. It would also be nice if you could run the pre-commit hooks so everything is formatted properly and address the errors in the new files (some minor things like missing docstring I think).

Also is it correct that I am getting DEBUG console output without changing any settings?

Furthermore, I am running into some problems running the plugin but maybe I am using it wrong. When I select the smac-pendigits run, the plugin does not seem to produce any explanations.

Server Log
```
SymbolicExplanations (DEBUG): Button pressed.
SymbolicExplanations (DEBUG): Enqueued SMAC-pendigits-015-0-0 (69666054159540351ed1f6fbf825f2cd).
RandomForestSurrogate (DEBUG): RandomForestSurrogate: Initial Seed = 0
RandomForestSurrogate (DEBUG): RandomForestSurrogate: Final Seed = 248400791
PDP (DEBUG): PDP: Initial Seed = 0
PDP (DEBUG): PDP: Final Seed = 1543538891
ICE (DEBUG): ICE: Initial Seed = True
ICE (DEBUG): ICE: Seed = Use existing
ICE (INFO): Recalculating ICE...
    |   Population Average    |             Best Individual              |
---- ------------------------- ------------------------------------------ ----------
 Gen   Length          Fitness   Length          Fitness      OOB Fitness  Time Left
   0    12.62          181.416       11         0.014005              N/A     17.39s
   1     8.73          20.5379       11         0.014005              N/A     14.85s
   2     7.70          0.39384       10         0.014005              N/A     12.92s
   3     4.63           0.2092       10         0.014005              N/A     11.07s
   4     2.15         0.086137        3        0.0140051              N/A      8.63s
   5     2.02        0.0345978        2        0.0140079              N/A      6.89s
   6     2.04        0.0389469        2        0.0140079              N/A      5.37s
   7     2.03        0.0395785        2        0.0140079              N/A      3.46s
   8     2.05        0.0408348        2        0.0140079              N/A      1.72s
   9     2.03         0.039717        2        0.0140079              N/A      0.00s
SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 for run_id 69666054159540351ed1f6fbf825f2cd.
SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 cached.
SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 deleted.
```

Bildschirmfoto 2024-02-09 um 11 51 20

File: Makefile (hunk 1 above)
Comment (sarah-segel) [reply Makefile:5]: Edited to 1.2 in Changelog.

File: deepcave/__init__.py (hunk 2 above)
Comment (sarah-segel) [reply deepcave/__init__.py:20]: Same as above.

File: docs/plugins/symbolic_explanations.rst (hunk 3 above)
Comment (sarah-segel) [reply docs/plugins/symbolic_explanations.rst]: Changed that in all cases of wrt

File: docs/plugins/symbolic_explanations.rst (hunk 4 above)
Comment (sarah-segel) [reply docs/plugins/symbolic_explanations.rst]: Added a sentence in the beginning.

File: deepcave/cli.py (hunk 5 above)
Comment (sarah-segel) [reply deepcave/cli.py]: Removed.

File: deepcave/config.py (hunk 6 above)
Comment (sarah-segel) [reply deepcave/config.py:54]: Not sure how they got there, removed :)

File: deepcave/utils/symbolic_regression.py (hunk 7 above)
Comment (sarah-segel) [reply deepcave/utils/symbolic_regression.py]: I unfortunately didn't track that and cannot find an example now, thus I just changed it to actually print the exception (so if I should see one at some point, I can edit it).

File: deepcave/utils/symbolic_regression.py (hunk 9 above)
Comment (sarah-segel) [reply deepcave/utils/symbolic_regression.py]: Added.

File: deepcave/plugins/hyperparameter/symbolic_explanations.py (hunk 10 above)
Comment (sarah-segel) [reply deepcave/plugins/hyperparameter/symbolic_explanations.py]: True, added.

**Comment by sarah-segel** [comment]
> Looks nice overall and the tests are working 💪 I have some minor remarks. It would also be nice if you could run the pre-commit hooks so everything is formatted properly and address the errors in the new files (some minor things like missing docstring I think).
> 
> Also is it correct that I am getting DEBUG console output without changing any settings?
> 
> Furthermore, I am running into some problems running the plugin but maybe I am using it wrong. When I select the smac-pendigits run, the plugin does not seem to produce any explanations.
> 
> Server Log
> 
> ```
> SymbolicExplanations (DEBUG): Button pressed.
> SymbolicExplanations (DEBUG): Enqueued SMAC-pendigits-015-0-0 (69666054159540351ed1f6fbf825f2cd).
> RandomForestSurrogate (DEBUG): RandomForestSurrogate: Initial Seed = 0
> RandomForestSurrogate (DEBUG): RandomForestSurrogate: Final Seed = 248400791
> PDP (DEBUG): PDP: Initial Seed = 0
> PDP (DEBUG): PDP: Final Seed = 1543538891
> ICE (DEBUG): ICE: Initial Seed = True
> ICE (DEBUG): ICE: Seed = Use existing
> ICE (INFO): Recalculating ICE...
>     |   Population Average    |             Best Individual              |
> ---- ------------------------- ------------------------------------------ ----------
>  Gen   Length          Fitness   Length          Fitness      OOB Fitness  Time Left
>    0    12.62          181.416       11         0.014005              N/A     17.39s
>    1     8.73          20.5379       11         0.014005              N/A     14.85s
>    2     7.70          0.39384       10         0.014005              N/A     12.92s
>    3     4.63           0.2092       10         0.014005              N/A     11.07s
>    4     2.15         0.086137        3        0.0140051              N/A      8.63s
>    5     2.02        0.0345978        2        0.0140079              N/A      6.89s
>    6     2.04        0.0389469        2        0.0140079              N/A      5.37s
>    7     2.03        0.0395785        2        0.0140079              N/A      3.46s
>    8     2.05        0.0408348        2        0.0140079              N/A      1.72s
>    9     2.03         0.039717        2        0.0140079              N/A      0.00s
> SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 for run_id 69666054159540351ed1f6fbf825f2cd.
> SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 cached.
> SymbolicExplanations (DEBUG): Job 69666054159540351ed1f6fbf825f2cd-48ad708a7a708995d8fdf94b34a2dca3 deleted.
> ```
> 
> Bildschirmfoto 2024-02-09 um 11 51 20

Regarding the debug logging, I changed the plugin (and other) specific logging settings in `logging.yml ` to `INFO`.

For the pre-commit hooks, I edited the files I added similar to the changes in the pre-commit branch.

Regarding the problem with the smac-pendigits run, this seems to be related to the problem I ran into with one of Daphne's runs (https://github.com/automl/DeepCAVE/issues/111). I added the example there.

@helegraf  would be great if you could re-check if everything is addressed :)

File: deepcave/utils/logging.yml [hunk 11]
```diff
@@ -22,6 +22,6 @@ loggers:
     handlers: [ console ]
     propagate: no
 root:
-  level: DEBUG
+  level: INFO
   handlers: [console]
-disable_existing_loggers: false
+disable_existing_loggers: true
```
Comment (alexandertornede) [inline deepcave/utils/logging.yml:27]: What is the reason for us setting this to true?

File: docs/plugins/configurations.rst [hunk 12]
```diff
@@ -13,7 +13,7 @@ Since configurations are used throughout the application, you might find links a
 plugin. This plugin is capable of answering following questions:
 
 * Where is the configuration coming from?
-* How are the objective values wrt the budgets?
+* How are the objective values with respect to the budgets?
```
Comment (alexandertornede) [inline docs/plugins/configurations.rst:16]: What are the ....

**Review by alexandertornede — APPROVED** [review]
Suggested some small adaptations that might be nice.

**Pull request merged**